lnwallet/chanfunding/coin_select.go

```diff
@@ -214,3 +214,57 @@ func CoinSelectSubtractFees(feeRate chainfee.SatPerKWeight, amt,
 
 	return selectedUtxos, outputAmt, changeAmt, nil
 }
+
+// CoinSelectUpToAmount attempts to select coins such that we'll spend up
+// to amt in total before fees, or all of the coins if not sufficient amount
+// available. The specified fee rate should be expressed in sat/kw for coin
+// selection to function properly.
+func CoinSelectUpToAmount(feeRate chainfee.SatPerKWeight, maxAmt, minAmt,
+	dustLimit btcutil.Amount, coins []Coin) ([]Coin, btcutil.Amount,
+	btcutil.Amount, error) {
+
+	// Get balance from coins.
+	var balance btcutil.Amount
+	for _, coin := range coins {
+		balance += btcutil.Amount(coin.Value)
+	}
+
+	// First we try to select coins up to the specified maxAmt.
+	outputAmt := maxAmt
+	selectedCoins, changeAmt, err := CoinSelect(
+		feeRate, maxAmt, coins,
+	)
+	if err != nil {
+
+		switch err.(type) {
+		// Without sufficient funds we spend it all by subtracting
+		// the fee directly from the output amount.
+		case *ErrInsufficientFunds:
+			selectedCoins, outputAmt, changeAmt, err = CoinSelectSubtractFees(
+				feeRate, balance, dustLimit, coins,
+			)
+			if err != nil {
+				return nil, 0, 0, err
+			}
+
+		default:
+			return nil, 0, 0, err
+		}
+	}
+
+	// Sanity check: Always make sure the output amount is below the
+	// specified max amount.
+	switch d := outputAmt - maxAmt; {
+	case d > dustLimit:
+		return nil, 0, 0, fmt.Errorf(
+			"non-negative difference in requested maximum funding "+
+				"amount %v and obtained amount %v", maxAmt,
+			outputAmt,
+		)
+
+	case d > 0:
+		outputAmt = maxAmt
```

Comment:
shouldn't this be an error?

Reply:
Since in the case `outputAmt != maxAmt` we already determined above that we don't have funds to fund `maxAmt`.

Reply:
In case `outputAmt != maxAmt` we would expect the routine to select coins below `maxAmt`. But due to the edge case (see https://github.com/lightningnetwork/lnd/pull/4029#issuecomment-592955663 and the corresponding test case https://github.com/lightningnetwork/lnd/pull/4029/files#diff-a19edac697b6152ce1c2fec4f6ac1d80R511), it is possible that coins are selected slightly above `maxAmt`. In case the difference is below the dust limit we set `outputAmt` to `maxAmt` and implicitly let the small difference go towards the fee, otherwise `outputAmt` is left unchanged.

Reply:
Aha, I see! Thanks for the explanation :)

Your fix does indeed seem to handle this edge case, but it feels a bit hacky. Could we instead try to fix the change output calculation in `CoinSelect`? (I see I already added a TODO there for this case).

Reply:
(preferably as its own independent PR with corresponding test cases)

Reply:
Yes, I think that's a good idea. I will look into that!

Reply:
(I am still looking into this. I just didn't find the time prior to now.)

Reply:
I stumbled over this again. Somehow this still makes me nervous. But reading this comment again and seeing the test case I see why this is needed. Perhaps just call the variable `overshoot` instead of `d` to make it a bit more clear?

Reply:
Agree on variable name change. It's a bit difficult to understand what's going on, so naming it `overshoot` would make it more clear.